keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1230 @@
+
+Booleans are rendered as `true` or `false`.
+
+Structs are rendered with curly-braces. Field names are not quoted unless they are not obviously strings (which includes prefixed label keys) or are one of the known type-ambiguous words (e.g. `no`).  Field values are rendered as per their type. Embedded structs will be rendered as inline fields.  Fields which have a `json` struct tag will be rendered according to that tag:
+  * the tag’s field name will be used
+  * if the tag’s name is “-”, the field will be skipped
+  * if the tag specifies “omitempty” it will be respected
+  * if the tag specifies “omitzero” it will be respected
+
+Maps are rendered with curly-braces. Keys are sorted and are not quoted unless they are type-ambiguous (e.g. `no`), and values are rendered as per their type.
+
+<<[UNRESOLVED quoted strings]>>
+Should we treat strings for field names and map keys:
+As described above, with label keys quoted but field names not
+Allow unquoted label keys and more complex (but not ambiguous) values
+Always quote strings, including ALL map keys and field names
+<<[/UNRESOLVED>>
+
+Slices and arrays are rendered with square-braces. Elements are rendered as per this list.
+
+Pointers are rendered as `null` or dereferenced and rendered as per the pointee’s type.
+
+Types which implement the Go standard `json.Marshaler` interface (`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an `any`, and then rendering that `any` as KYAML.
+
+Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are rendered as per their type.  `MarshalYAML()` is not a Go standard and is not used by KYAML.
+
+<<[UNRESOLVED yaml self-marshaling]>>
+Should we do something for YAML self-marshalling?  It is non-standard and we do not use it in Kubernetes.
+<<[/UNRESOLVED>>
+
+Interfaces are rendered as `null` or their true value type.
+
+Any other type (as per `reflect.Value.Kind()`) will produce an error.
+
+### Multi-line strings
+
+For better or worse, multi-line strings are a thing that people do in Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how we want to handle it.  Consider these examples of conventional YAML.
+
+Example 1:
+```
+example_1: |
+    This
+    is a
+    multi-line string
+```
+
+Example 2:
+```
+example_2: |
+    this:
+      is:
+      - embedded
+      - yaml
+      where:
+        whitespace:
+          matters: true
+```
+
+This style is not available in YAML flow-style.
+
+#### Option 1. Treat it like a Go quoted-and-escaped string
+
+This always comes out as one line.
+
+```
+example_1: “This\nis a\nmulti-line string\n”
+example_2: "this:\n  is:\n  - embedded\n  - yaml\n  where:\n    whitespace:\n      matters: true\n"
+```
+
+Pro: Easy and well-defined
+Pro: Can be pasted into most C-like languages to get the original string.
+Con: Hard to read and write by humans when it gets long.
```

Comment:
I agree that this is not an option, in particular with an eye towards using KYAML for logging and test output. The original klog structured text log output used this and we promptly got bug reports and reverts because it made the log output less readable. I then added multi-line output.

The problem is that none of the options below seem particularly appealing either :disappointed:

Reply:
I think 2) is tolerable. All multi-line strings are a bit awkward in yaml, the "flow" style strings have different problems we didn't discuss here because they can't be used with block style.

Reply:
I agree that option 2 is okayish. That `this:` and `  is:` in example 2 are shifted by one character (the `\` before `  is:`) is a pity.

BTW, you seem to use the opposite definition of "flow" and "block" than the KEP? Option 1 is said to be "not available in YAML flow-style" with the string using block style, so options 2 onward must be flow-style strings. Or am I mixing things up? I hadn't seen this terminology before.

Reply:
flow-style is with `{}` and `[]`

block-style is with whitespace

https://yaml.org/spec/1.2.2/

Section 2.1 says:

"""
YAML also has flow styles, using explicit indicators rather than indentation to denote scope. The flow sequence is written as a comma separated list within square brackets. In a similar manner, the flow mapping uses curly braces.
"""

Section 3.2.3.1 says:

"""
There are two groups of styles. Block styles use indentation to denote structure. In contrast, flow styles rely on explicit indicators.
"""

If we got that wrong anywere, please flag it :)

Reply:
The KEP has it right, but Ben seems to use it the other way around in his comments - or I am misreading them.

Reply:
[I transposed flow and block in the comment above, I will edit it]

> I agree that option 2 is okayish. That this: and   is: in example 2 are shifted by one character (the \ before   is:) is a pity.

Yeah, though if you're depending on significant leading whitespace *inside of a yaml multi-line string* there's simply no non-error prone format for that particular aspect.

Some of them require less characters, but they're not any less confusing or error prone.

In general if you want to guarantee leading whitespace you need to escape it.

Reply:
> That this: and   is: in example 2 are shifted by one character (the \ before   is:) is a pity.

Variation B of option 2 addresses this nicely - it's my favorite solution.

Reply:
Agreed.